allauth/account/views.py

```diff
@@ -602,7 +607,19 @@ def form_valid(self, form):
         return super(PasswordResetFromKeyView, self).form_valid(form)
 
     def _response_bad_token(self, request, uidb36, key, **kwargs):
-        return self.render_to_response(self.get_context_data(token_fail=True))
+        # Special-casing ajax/non-ajax scenarios due to #890
+        if request.is_ajax():
+            data = {
+                'errors': ['token_failed'],
+            }
+            response = _ajax_response(request,
+                                      HttpResponseBadRequest(reason=data))
```

Comment:
For HttpResponse, the reason is supposed to be a string type. This is not enforced, so the above code will work, but it may break in the future. Also, allauth so far managed to stay away from being opinionated with respect to the contents of the XHR responses. This would be the first place (adapter excluded) where choices are being made as to what content is being passed.  So I am wondering, wouldn't it be better to simpy raise a 400 without passing any specific data?

Perhaps cleaner is this approach: we could create a specific form (`UserTokenForm`) and move the token validation and user lookup in there. Then, in PasswordResetFromKeyView instantiate this form (not passing request.GET/POST but simply the uidb36/key in directly). If the validation fails, then this form could be passed to the `ajax_response`, where in the adapter people can do whathever they please, including a data error response like above.